### PR TITLE
add set function to arraylist

### PR DESCRIPTION
**Pull request by clownpriest**
so you can set a value without growing the underlying buffer (like
insert function), maybe better than working directly with self.items

### CODE REVIEW

**Comment by BraedonWooding** [comment]
Thanks for opening a PR :)!  A few things however,

Since you don't actually increase the `.len` field in the structure when iterating (using the iterator) through it you won't actually iterator across the items you set.  Maybe perhaps you meant to compare against the `n > self.len` which would guarantee that you are only accessing the current 'array'?  Since going outside that is really against the idea of an array list.  If you want to set items anywhere in that internal array you are right probably better to go with `self.items` directly as that is more explicit.

Also `l` should be called `self` by convention.

**Comment by clownpriest** [comment]
i used "l" rather than "self" just because all the other function currently use "l". latest commit compares bounds with l.len rather than items.len. i'm fine to close this PR and just set explicitly through self.items[], though then the caller would have to worry about iterators and self.len, no? maybe it's a non-issue

**Comment by thejoshwolfe** [comment]
Please add a test for your new api.

Adding tests would help catch bugs like the one @BraedonWooding pointed out. Tests are also a nice way to document the usage of your code. In general tests are good to have and will make your code easier to review and more attractive to merge.

**Comment by clownpriest** [comment]
thank you for the recommendation! found a bug in the process of adding tests. latest commit has tests and the bug fix

**Comment by BraedonWooding** [comment]
@clownpriest sometimes it errors out like that, often it'll work when re-run (note: how under the details for travis-ci/pr it says it '!' error'd not failed), you can force a re-run by closing PR and reopening it or someone with admin rights I think can re-run it :).

File: std/array_list.zig [hunk 1]
```diff
@@ -200,6 +210,20 @@ test "basic ArrayList test" {
...
     list.appendSlice([]const i32{}) catch unreachable;
     assert(list.len == 9);
+    
+    // can only set on indices < self.len
+    list.set(7, 33) catch unreachable;
+    list.set(8, 42) catch unreachable;
+    
+    list.set(9, 99) catch |err| {
+        assert(err == error.OutOfBounds);
+    };
```
[the remark this answers is not in the document]
Comment (andrewrk) [reply std/array_list.zig]: I would recommend `@panic("assertion failed")` over `assert(false)`.

**Comment by andrewrk** [comment]
merged in e0092ee4a573502a4110a6d4aeb7e7d3cdc8987b

**Pull request closed without merging**